### PR TITLE
Delete elements verification

### PR DESCRIPTION
**Pull request by swilly22**
Both UNWIND and DELETE should be operating on Arithmetic expressions,
Consider:
`MATHC p = ()-[]->() UNWIND nodes(p) AS node DELETE node`
`MATHC p = () DELETE nodes(p)[0]`

### CODE REVIEW

File: src/arithmetic/numeric_funcs/numeric_funcs.c [hunk 1]
```diff
@@ -120,6 +121,30 @@ SIValue AR_SIGN(SIValue *argv, int argc) {
...
+		return arg;
+	case T_DOUBLE:
+		// Remove floating point.
+		return SI_LongVal(floor(arg.doubleval));
+	case T_STRING:
+		errno = 0;
+		double parsedval = strtod(arg.stringval, &sEnd);
+		/* The input was not a complete number or represented a number that
+		 * cannot be represented as a double.
+		 * Create a string SIValue. */
```
Comment (jeffreylovitz) [inline src/arithmetic/numeric_funcs/numeric_funcs.c]: Delete last line of comment.

File: src/arithmetic/numeric_funcs/numeric_funcs.c [hunk 2]
```diff
@@ -120,6 +121,30 @@ SIValue AR_SIGN(SIValue *argv, int argc) {
...
+		if(sEnd[0] != '\0' || errno == ERANGE) return SI_NullVal();
+		// Remove floating point.
+		return SI_LongVal(floor(parsedval));
+	default:
```
Comment (jeffreylovitz) [inline src/arithmetic/numeric_funcs/numeric_funcs.c]: This function should also handle NULL values, with NULL inputs returning null SIValues. This is important for cases like operating on a property that may not exist on all inputs.

File: src/arithmetic/numeric_funcs/numeric_funcs.c [hunk 3]
```diff
@@ -172,4 +197,9 @@ void Register_NumericFuncs() {
 	types = array_append(types, (SI_NUMERIC | T_NULL));
 	func_desc = AR_FuncDescNew("sign", AR_SIGN, 1, 1, types, true);
 	AR_RegFunc(func_desc);
+
+	types = array_new(SIType, 1);
+	types = array_append(types, (SI_NUMERIC | T_STRING));
```
Comment (jeffreylovitz) [inline src/arithmetic/numeric_funcs/numeric_funcs.c]: With the change at 143, `SI_NUMERIC | T_STRING | T_NULL`

File: src/ast/ast_build_op_contexts.c [hunk 4]
```diff
@@ -89,35 +89,17 @@ EntityUpdateEvalCtx *AST_PrepareUpdateOp(const cypher_astnode_t *set_clause) {
...
 		const cypher_astnode_t *ast_expr = cypher_ast_delete_get_expression(delete_clause, i);
-		assert(cypher_astnode_type(ast_expr) == CYPHER_AST_IDENTIFIER);
-
-		const char *ast_alias = cypher_ast_identifier_get_name(ast_expr);
-		// Retrieve the canonical pointer for the QueryGraph entity.
-		char *alias = raxFind(ast->canonical_entity_names, (unsigned char *)ast_alias, strlen(ast_alias));
-
-		/* We need to determine whether each alias refers to a node or edge.
-		 * Currently, we'll do this by consulting with the QueryGraph. */
-		EntityType type = QueryGraph_GetEntityTypeByAlias(qg, alias);
```
Comment (jeffreylovitz) [inline src/ast/ast_build_op_contexts.c]: I'm really happy to see this logic gone!

File: src/ast/ast_validations.c [hunk 5]
```diff
@@ -883,24 +873,6 @@ static AST_Validation _Validate_UNWIND_Clauses(const AST *ast, char **reason) {
...
-				res = AST_INVALID;
-				goto cleanup;
-			}
-		} else if(!(type == CYPHER_AST_COLLECTION || type == CYPHER_AST_IDENTIFIER ||
-					type == CYPHER_AST_PROPERTY_OPERATOR)) {
-			asprintf(reason, "UNWIND expects a list argument or an list identifier; encountered ''%s'",
-					 cypher_astnode_typestr(type));
-			res = AST_INVALID;
-			goto cleanup;
-		}
```
Comment (jeffreylovitz) [inline src/ast/ast_validations.c]: Ideally we'd be able to capture things like `UNWIND 5 AS a RETURN a` as syntax errors, though I'd be fine with a TODO for the moment.

We should add some run-time captures in `UnwindConsume`, however. While this example just returns an empty result set, `UNWIND toString(5) AS a RETURN a` causes a crash for me.

File: src/execution_plan/ops/op_delete.c [hunk 6]
```diff
@@ -88,15 +68,30 @@ static Record DeleteConsume(OpBase *opBase) {
...
-		Edge *e = Record_GetEdge(r, op->edges_to_delete[i]);
-		op->deleted_edges = array_append(op->deleted_edges, *e);
+	/* Expression should be evaluated to either a node or an edge
+	 * which will be marked for deletion, if an expression is evaluated
+	 * to a different value type e.g. Numeric a run-time expection is thrown. */
+	for(int i = 0; i < op->exp_count; i++) {
+		AR_ExpNode *exp = op->exps[i];
+		SIValue value = AR_EXP_Evaluate(exp, r);
+
+		if(!(SI_TYPE(value) & SI_GRAPHENTITY)) {
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_delete.c]: This logic could be put in the `else` block just below, but I don't have a strong preference if you think this is cleaner.

File: src/execution_plan/ops/op_delete.c [hunk 7]
```diff
@@ -88,15 +68,30 @@ static Record DeleteConsume(OpBase *opBase) {
...
+			char *error;
+			asprintf(&error, "Delete type mismatch, expecting either Node or Relationship.");
+			QueryCtx_SetError(error);
+			QueryCtx_RaiseRuntimeException();
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_delete.c]: If we encounter an error, I think we should clear the `deleted_nodes` and `deleted_edges` arrays now, or else the deletes will still be committed in the free routine.

File: src/execution_plan/ops/op_unwind.c [hunk 8]
```diff
@@ -117,4 +117,3 @@ static void UnwindFree(OpBase *ctx) {
 		op->currentRecord = NULL;
 	}
 }
-
```
Comment (jeffreylovitz) [inline src/execution_plan/ops/op_unwind.c]: What is our style policy on trailing newlines? My autoformatter is adding them, should I change that?

File: src/arithmetic/numeric_funcs/numeric_funcs.c (hunk 1 above)
Comment (swilly22) [reply src/arithmetic/numeric_funcs/numeric_funcs.c]: 👍

File: src/arithmetic/numeric_funcs/numeric_funcs.c (hunk 2 above)
Comment (swilly22) [reply src/arithmetic/numeric_funcs/numeric_funcs.c]: 👍

File: src/arithmetic/numeric_funcs/numeric_funcs.c (hunk 3 above)
Comment (swilly22) [reply src/arithmetic/numeric_funcs/numeric_funcs.c]: 👍

File: src/ast/ast_validations.c (hunk 5 above)
Comment (swilly22) [reply src/ast/ast_validations.c]: 👍 Fixed

File: src/execution_plan/ops/op_delete.c (hunk 6 above)
Comment (swilly22) [reply src/execution_plan/ops/op_delete.c]: Absolutely!

File: src/execution_plan/ops/op_delete.c (hunk 7 above)
Comment (swilly22) [reply src/execution_plan/ops/op_delete.c]: Good catch!

File: src/execution_plan/ops/op_unwind.c (hunk 8 above)
Comment (swilly22) [reply src/execution_plan/ops/op_unwind.c]: A single newline at the end of each file

**Pull request merged**